src/Parallel/Algorithm.hpp

```diff
@@ -182,6 +175,28 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>
   /// Charm++ migration constructor, used after a chare is migrated
   constexpr explicit AlgorithmImpl(CkMigrateMessage* /*msg*/) noexcept;
 
+  void pup(PUP::er& p) noexcept override {  // NOLINT
```

Comment:
Is there some way we can test this? For example in the `Test_Algorithm*` executables?

Reply:
So, it's a little tricky to test the serialization manually in those tests because the algorithm tests are actually run in the charm runtime.
What we could do easily is enable the CommonLBs compiler flag (either for those algorithm tests, or just generally), and have an algorithm test run with e.g. `RandCentLB` that will always migrate, and confirm that the output has evidence of migration from LBDebug.

Reply:
A checkpoint-restart based case might be nicer, in that it could test the different kinds of chares? Not helpful now, but could be added in a few PRs after this one.

Reply:
Would it not be possible to grab the `local()` in an action and then serialize/deserialize and test everything in the deserialized is correct or would this confuse Charm++? Or would having a test where you construct just an AlgorithmImpl and serialize/deserialize work and be easier? Generally when we've had bugs they've been subtly hidden in `pup` functions and this seems like one of the more important pup functions :P

Reply:
Right, so I think we could do that (I don't know if we've ever manually serialized an object that inherits from `CBase_blahblah` in a test, but I don't see anything going immediately wrong) -- the main challenge is then how we examine all of the things that we claim to be `pup`ing.
I agree this needs to be tested well because bugs in this will be subtle and affect effectively every executable. 

We could create an equivalence operator that verifies each of the member variables are the same before and after pupping. This is not terribly convincing, because all it really tests is that we forgot the same member variables in both the pup and the operator==...

Ideally, we'd have a test where the packed and unpacked algorithm is actually used to run the rest of the test, but I'm not sure how to do that without going through the charm runtime. Eventually, we'll need to have the ability to dynamically add and remove elements, but our phase structure currently can't do that I think.

We could add utilities to expose manual migration to the algorithm if that's more trustworthy than a balancer, then confirm that the element has actually been moved to a new processor. The one part of that version or the balancer version that's a bit unclear to me is whether it is guaranteed that charm will actually universally call the pup function when moving between cores on the same node. We could verify that last point by adding an extra debugging member function that returns the current memory address of the databox or something to verify that data has actually been moved.

What do you think?